### DISCUSSION TITLE
[Bug] Value interpolation between different config files

### DISCUSSION THREAD

**Discussion by apoorvagnihotri** [facebookresearch/hydra#2288] [Q&A] [answered]
# 🐛 Bug
## Description
Value interpolation doesn't seem to be functioning when trying to push a value in parent config to the child config.

## Checklist
- [x] I checked on the latest version of Hydra
- [x] I created a minimal repro (See [this](https://stackoverflow.com/help/minimal-reproducible-example) for tips).

## To reproduce
In order to recreate the issue, you need 4 files. The idea is to run, `run.py` which is basically equivalent to a pipeline, and any arguments that are passed to `run.py`, like `debug`, should be passed down to `preprocesses.py`.

```python
# preprocesses.py

import hydra
from omegaconf import DictConfig, OmegaConf


@hydra.main(version_base=None, config_path="./conf/preprocess", config_name="default")
def preprocess(cfg: DictConfig):
    print ("Preprocessing Starting", end="\n\n")
    print (OmegaConf.to_yaml(cfg))
    print ("Preprocessing Done")

if __name__ == "__main__":
    preprocess()
```

```python
# run.py

import hydra
from omegaconf import DictConfig, OmegaConf
from preprocesses import preprocess


@hydra.main(version_base = None, config_path="./conf", config_name="conf")
def run(cfg: DictConfig):
    print ("Run Starting")
    print (OmegaConf.to_yaml(cfg))
    preprocess(cfg.preprocess)
    print ("Run Done")

if __name__ == "__main__":
    run()
```

```yaml
# conf/preprocess/default.yaml

db:
  user: omry
  debug: false
```

```yaml
# conf/main.yaml

defaults:
  - preprocess: default
  - _self_

application:
  version: 1.0.0
  debug: true

preprocess:
  db:
    debug: ${application.debug}
```

** Output **

When running the command `python .\run.py --config-name=main`, I get the following output.
```
Run Starting
preprocess:
  db:
    user: omry
    debug: ${application.debug}
application:
  version: 1.0.0
  debug: true

Preprocessing Starting

db:
  user: omry
  debug: ${application.debug}

Preprocessing Done
Run Done
```

## Expected Behavior
I expected the value interpolation (as shown [here](https://hydra.cc/docs/tutorials/basic/your_first_app/using_config/)) to happen. Expected output:

```
Run Starting
preprocess:
  db:
    user: omry
    debug: true
application:
  version: 1.0.0
  debug: true

Preprocessing Starting

db:
  user: omry
  debug: true

Preprocessing Done
Run Done
```

## System information
- **Hydra Version** :  1.2.0
- **Python version** :  3.10.5
- **Virtual environment type and version** :  conda (4.12.0)
- **Operating system** :  windows 11

**Comment by Jasha10** [comment, marked as answer]
Interpolations are resolved lazily on access. For example:
```python
>>> cfg = OmegaConf.create("""
... application: {debug: true}
... interp: ${application.debug}
... """)
>>> print(OmegaConf.to_yaml(cfg))
application:
  debug: true
interp: ${application.debug}
>>> cfg.interp  # lazy resolution happens here:
True
```
If you want to print the resolved values, you can pass keyword argument `resolve=True` to the `OmegaConf.to_yaml` function:
```python
>>> print(OmegaConf.to_yaml(cfg, resolve=True))
application:
  debug: true
interp: true
```
Alternatively, you can call [`OmegaConf.resolve`](https://omegaconf.readthedocs.io/en/2.1_branch/usage.html#omegaconf-resolve). This will modify the config in-place, replacing the interpolations with their resolved values:
```python
>>> OmegaConf.resolve(cfg)
>>> print(OmegaConf.to_yaml(cfg))
application:
  debug: true
interp: true
```

For more info, see the OmegaConf docs on [Variable Interpolation](https://omegaconf.readthedocs.io/en/2.1_branch/usage.html#variable-interpolation).
